File: frontend/src/calc.ts
/**
 * Mirror of internal/calc, used for live display while the user types. The Go
 * side re-derives the same values for the PDF, so the two must stay in step.
 */

export type RowLike = {
  cantitate: number;
  pretFaraTva: number;
  pretCuTva: number;
};

export type Totals = {
  cantitate: number;
  valoareFaraTva: number;
  valoareCuTva: number;
};

/** Rounds to two decimals, half away from zero. */
export function round2(value: number): number {
  const rounded = Math.round(Math.abs(value) * 100) / 100;
  return value < 0 ? -rounded : rounded;
}

/** cantitate x pret, rounded to two decimals. */
export function valoare(cantitate: number, pret: number): number {
  return round2(cantitate * pret);
}

/** Sums a table's numeric columns. */
export function totals(rows: RowLike[]): Totals {
  const sum = rows.reduce<Totals>(
    (acc, row) => ({
      cantitate: acc.cantitate + row.cantitate,
      valoareFaraTva: acc.valoareFaraTva + valoare(row.cantitate, row.pretFaraTva),
      valoareCuTva: acc.valoareCuTva + valoare(row.cantitate, row.pretCuTva),
    }),
    { cantitate: 0, valoareFaraTva: 0, valoareCuTva: 0 },
  );
  return {
    cantitate: round2(sum.cantitate),
    valoareFaraTva: round2(sum.valoareFaraTva),
    valoareCuTva: round2(sum.valoareCuTva),
  };
}

/** The signed gap between the two tables' "valoare cu TVA" totals. */
export function diferenta(
  iesireCuTva: number,
  intrareCuTva: number,
): { tip: 'plus' | 'minus' | ''; valoare: number } {
  const delta = round2(iesireCuTva - intrareCuTva);
  if (delta > 0) return { tip: 'plus', valoare: delta };
  if (delta < 0) return { tip: 'minus', valoare: -delta };
  return { tip: '', valoare: 0 };
}

/** The same magnitude, labelled the way the bottom-right block labels it. */
export function incarcaDescarca(
  iesireCuTva: number,
  intrareCuTva: number,
): { tip: 'incarca' | 'descarca' | ''; valoare: number } {
  const { tip, valoare: value } = diferenta(iesireCuTva, intrareCuTva);
  if (tip === 'plus') return { tip: 'incarca', valoare: value };
  if (tip === 'minus') return { tip: 'descarca', valoare: value };
  return { tip: '', valoare: 0 };
}

/**
 * The profit margin of the butchering, as a percentage of what went in:
 * (total iesire − total intrare) / total intrare × 100, both "cu TVA" so the
 * figure agrees with Diferență. Negative when the butchering lost value.
 *
 * Returns undefined — not Infinity or NaN — when nothing went in, since a
 * margin on a zero cost has no meaning; callers render that as a blank.
 *
 * Unlike the rest of this module this has no counterpart in internal/calc: it
 * is a working figure shown on screen only, never stored and never printed.
 */
export function marjaProfit(iesireCuTva: number, intrareCuTva: number): number | undefined {
  if (intrareCuTva === 0) return undefined;
  return round2(((iesireCuTva - intrareCuTva) / intrareCuTva) * 100);
}

/**
 * The multiplier a TVA rate applies to a price without TVA, or undefined when
 * the rate cannot be applied at all: at -100% or below the multiplier is zero
 * or negative, which has no usable inverse. Callers treat undefined as "leave
 * the other price alone" rather than writing a zero or an Infinity.
 */
function multiplicatorTva(cota: number): number | undefined {
  const factor = 1 + cota / 100;
  return factor > 0 ? factor : undefined;
}

/** The price with TVA that a price without TVA implies at the given rate. */
export function pretCuTvaDin(pretFaraTva: number, cota: number): number | undefined {
  const factor = multiplicatorTva(cota);
  return factor === undefined ? undefined : round2(pretFaraTva * factor);
}

/** The price without TVA that a price with TVA implies at the given rate. */
export function pretFaraTvaDin(pretCuTva: number, cota: number): number | undefined {
  const factor = multiplicatorTva(cota);
  return factor === undefined ? undefined : round2(pretCuTva / factor);
}

File: internal/calc/calc.go
// Package calc holds the arithmetic of the proces verbal: per-row values,
// per-table totals and the difference between the two tables.
package calc

import (
	"math"

	"proces-verbal-transare/internal/model"
)

// Totals is the sum of a table's numeric columns.
type Totals struct {
	Cantitate      float64 `json:"cantitate"`
	ValoareFaraTVA float64 `json:"valoareFaraTva"`
	ValoareCuTVA   float64 `json:"valoareCuTva"`
}

// Round2 rounds to two decimals, half away from zero.
func Round2(v float64) float64 {
	r := math.Round(math.Abs(v)*100) / 100
	if v < 0 {
		return -r
	}
	return r
}

// Round3 rounds to three decimals, half away from zero. Ratios of the carcass
// are kept at this precision rather than two: at two decimals one hundredth of
// a percent of a 162.2 Kg carcass is 16 grams, enough to shift a row's rounded
// quantity, and the ratios must give back the quantities they were derived
// from.
func Round3(v float64) float64 {
	r := math.Round(math.Abs(v)*1000) / 1000
	if v < 0 {
		return -r
	}
	return r
}

// ValoareFaraTVA is cantitate x pret fara TVA.
func ValoareFaraTVA(cantitate, pret float64) float64 {
	return Round2(cantitate * pret)
}

// ValoareCuTVA is cantitate x pret cu TVA.
func ValoareCuTVA(cantitate, pret float64) float64 {
	return Round2(cantitate * pret)
}

// TotalsIntrare sums the "ce intra" table.
func TotalsIntrare(rows []model.IntrareRow) Totals {
	var t Totals
	for _, r := range rows {
		t.Cantitate += r.Cantitate
		t.ValoareFaraTVA += ValoareFaraTVA(r.Cantitate, r.PretFaraTVA)
		t.ValoareCuTVA += ValoareCuTVA(r.Cantitate, r.PretCuTVA)
	}
	return roundTotals(t)
}

// TotalsIesire sums the "ce iese" table.
func TotalsIesire(rows []model.IesireRow) Totals {
	var t Totals
	for _, r := range rows {
		t.Cantitate += r.Cantitate
		t.ValoareFaraTVA += ValoareFaraTVA(r.Cantitate, r.PretFaraTVA)
		t.ValoareCuTVA += ValoareCuTVA(r.Cantitate, r.PretCuTVA)
	}
	return roundTotals(t)
}

func roundTotals(t Totals) Totals {
	return Totals{
		Cantitate:      Round2(t.Cantitate),
		ValoareFaraTVA: Round2(t.ValoareFaraTVA),
		ValoareCuTVA:   Round2(t.ValoareCuTVA),
	}
}

// Diferenta reports the signed gap between the two tables' "valoare cu TVA"
// totals as a ("plus"|"minus"|"") tip and a non-negative magnitude.
func Diferenta(totalIesireCuTVA, totalIntrareCuTVA float64) (string, float64) {
	d := Round2(totalIesireCuTVA - totalIntrareCuTVA)
	switch {
	case d > 0:
		return "plus", d
	case d < 0:
		return "minus", -d
	default:
		return "", 0
	}
}

// IncarcaDescarca is the same magnitude as Diferenta, labelled the way the
// bottom-right block of the form labels it.
func IncarcaDescarca(totalIesireCuTVA, totalIntrareCuTVA float64) (string, float64) {
	tip, val := Diferenta(totalIesireCuTVA, totalIntrareCuTVA)
	switch tip {
	case "plus":
		return "incarca", val
	case "minus":
		return "descarca", val
	default:
		return "", 0
	}
}

// PretFaraTVA is the price without TVA that a price with TVA implies at the
// given rate, in percent. The second result is false when the rate cannot be
// applied at all — at -100% or below the multiplier is zero or negative, which
// has no usable inverse — and callers should then leave the price alone rather
// than store an infinity.
//
// Its counterpart in the frontend is pretFaraTvaDin in frontend/src/calc.ts,
// which must stay in step with it. Only this direction exists on the Go side:
// deriving a price with TVA is something only the form does, while the Go side
// needs the reverse to prefill a new document's "ce iese" rows from the stored
// product prices (see App.NewDocumentDraft).
func PretFaraTVA(pretCuTVA, cota float64) (float64, bool) {
	factor := 1 + cota/100
	if factor <= 0 {
		return 0, false
	}
	return Round2(pretCuTVA / factor), true
}
